google-cloud-clients/google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java

```diff
@@ -1122,11 +1124,50 @@ public static SignUrlOption signWith(ServiceAccountSigner signer) {
 
     /**
      * Use a different host name than the default host name 'https://storage.googleapis.com'. This
-     * must also include the scheme component of the URI.
+     * must also include the scheme component of the URI. Note that this cannot be used alongside
+     * {@code withVirtualHostName()}.
      */
     public static SignUrlOption withHostName(String hostName) {
       return new SignUrlOption(Option.HOST_NAME, hostName);
     }
+
+    /**
+     * Use a virtual hosted-style hostname, which includes the bucket in the host portion of the URI
+     * rather than the path, e.g. 'https://mybucket.storage.googleapis.com'. This must also include
+     * the scheme component of the URI. Note that this cannot be used alongside {@code
+     * withHostName()}. For V4 signing, this also sets the "host" header in the canonicalized
+     * extension headers to the specified value, minus the "http[s]://", unless that header is
+     * supplied via the {@code withExtHeaders()} method.
+     *
+     * @see <a href="https://cloud.google.com/storage/docs/request-endpoints">Request Endpoints</a>
+     */
+    public static SignUrlOption withVirtualHostName(String virtualHostName) {
```

Comment:
We probably don't need it unless someone asks; I was mostly trying to plan ahead in case anyone needed support for CNAME-based signed URLs.  I went ahead and removed it for now.

Reply:
ah gotcha, what about `withHostName()`? Doesn't this cover this case?

Reply:
So, I thought `withHostName()` only was useful for devs, e.g. sending requests to staging/test endpoints... meaning this would be interchangeable with anywhere you'd use `storage.googleapis.com` in a URL.  But looking at https://github.com/googleapis/google-cloud-java/issues/3140, it seems this was written to be used with CDNs.

But, what's confusing is... from all the info I can find (our docs, along with external CDN docs) on CNAME'ing and using CDNs with GCS, you're supposed to make the CDN or CNAME record to point at `<bucket_name>.storage.googleapis.com` (virtual hosted-style URL)... but from reading the code, it's producing path-style URLs (`https://myurl.somealias.com/this-bucket-name-is-in-the-path/object-name`).

From a semantics standpoint, this *should* be used in the manner I mentioned at the top.  However, I'm not actually sure how it's being used by the user(s) that requested this functionality.  I'll comment and ask on the issue they filed, but I'm not sure how likely I am to get a reply :upside_down_face:

Reply:
From looking at the tests again, when using `withHostName(String)`, we're definitely expecting to replace the normal hostname and for the bucket to be in the path portion of the URL; this applies for scenarios where we're using `withHostName()` and generating a path-style, V2-signed URL.  None of the other scenarios are currently tested.  That being said, I've done manual testing with other hostnames (i.e. against our test environment) and am confident that what I've got coded up is correct behavior.  If users need a different host header to be used (but note that from all cases I can find, this would cause the signature to become invalid), they can override it using the `withExtHeaders()` method.

- For V2 and V4 path-style URLs, the hostname should change, and the bucket should be in the path.
  - It looks like we don't even bother signing the "host" header for V2 URLs.  But for V4 URLs, the "host" header **should** be set to the new hostname.  We haven't been doing that, but I just added a fix for that in another commit.  This is another option combination that we need a test for.
- For V2 and V4 virtual hosted-style URLs, the hostname should have the bucket name (plus a period), as supplied in the `BlobInfo`, prepended to it, and the bucket should not appear in the path.  We currently have no tests for virtual hosted-style URLs, but I've manually tested them with V2 and V4 signing, both with and without an alternative hostname (I used our test environment and its hostname for testing).
  - Again, V2 URLs don't care about the host since we don't sign it. But V4 URLs should set the host to the virtual hosted-style hostname; I've added a fix to do this.

Reply:
The "with overloads" part is currently incorrect; what I had before (supplying a hostname string arg to the `withVirtualHostedStyle` method) would be the correct way to do the "with overloads" part, since the bucket name **is** the hostname, and you wouldn't want to prepend the bucket name **again**, e.g. ending up with `subdomain.domain.tld.subdomain.domain.tld` instead of just `subdomain.domain.tld` for your hostname.

I took out the overload in the last couple commits; the virtual hosted style currently assumes you won't be using a CNAME hostname. But, we can add the overload in a subsequent commit.  Seems worth breaking that functionality out into its own PR.  Thoughts?

Reply:
Acking we discussed this yesterday and splitting off overloads into a separate request is a follow-up action.